File: src/main/scala/mffs/item/mode/ItemModePyramid.java
package mffs.item.mode;

import resonant.api.mffs.IFieldInteraction;
import resonant.api.mffs.IProjector;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.tileentity.TileEntity;
import org.lwjgl.opengl.GL11;
import resonant.lib.prefab.vector.Cuboid;
import universalelectricity.core.transform.vector.Vector3;

import java.util.HashSet;
import java.util.Set;

public class ItemModePyramid extends ItemMode
{
	public ItemModePyramid(int i)
	{
		super(i, "modePyramid");
	}

	@Override
	public Set<Vector3> getExteriorPoints(IFieldInteraction projector)
	{
		final Set<Vector3> fieldBlocks = new HashSet<Vector3>();

		final Vector3 posScale = projector.getPositiveScale();
		final Vector3 negScale = projector.getNegativeScale();

		final int xStretch = posScale.xi() + negScale.xi();
		final int yStretch = posScale.yi() + negScale.yi();
		final int zStretch = posScale.zi() + negScale.zi();
		final Vector3 translation = new Vector3(0, -negScale.yi(), 0);

		final int inverseThickness = (int) Math.max((yStretch + zStretch) / 4f, 1);
		System.out.println(inverseThickness);

		for (float y = 0; y <= yStretch; y += 1f)
		{
			for (float x = -xStretch; x <= xStretch; x += 1f)
			{
				for (float z = -zStretch; z <= zStretch; z += 1f)
				{
					double yTest = (y / yStretch) * inverseThickness;
					double xzPositivePlane = (1 - (x / xStretch) - (z / zStretch)) * inverseThickness;
					double xzNegativePlane = (1 + (x / xStretch) - (z / zStretch)) * inverseThickness;

					// Positive Positive Plane
					if (x >= 0 && z >= 0 && Math.round(xzPositivePlane) == Math.round(yTest))
					{
						fieldBlocks.add(new Vector3(x, y, z).add(translation));
						fieldBlocks.add(new Vector3(x, y, -z).add(translation));
					}

					// Negative Positive Plane
					if (x <= 0 && z >= 0 && Math.round(xzNegativePlane) == Math.round(yTest))
					{
						fieldBlocks.add(new Vector3(x, y, -z).add(translation));
						fieldBlocks.add(new Vector3(x, y, z).add(translation));
					}

					// Ground Level Plane
					if (y == 0 && (Math.abs(x) + Math.abs(z)) < (xStretch + yStretch) / 2)
					{
						fieldBlocks.add(new Vector3(x, y, z).add(translation));
					}
				}
			}
		}

		return fieldBlocks;
	}

	@Override
	public Set<Vector3> getInteriorPoints(IFieldInteraction projector)
	{
		Set<Vector3> fieldBlocks = new HashSet<Vector3>();

		Vector3 posScale = projector.getPositiveScale();
		Vector3 negScale = projector.getNegativeScale();

		int xStretch = posScale.xi() + negScale.xi();
		int yStretch = posScale.yi() + negScale.yi();
		int zStretch = posScale.zi() + negScale.zi();
		Vector3 translation = new Vector3(0, -0.4, 0);

		for (float x = -xStretch; x <= xStretch; x++)
		{
			for (float z = -zStretch; z <= zStretch; z++)
			{
				for (float y = 0; y <= yStretch; y++)
				{
					Vector3 position = new Vector3(x, y, z).add(translation);

					if (this.isInField(projector, Vector3.translate(position, new Vector3((TileEntity) projector))))
					{
						fieldBlocks.add(position);
					}
				}
			}
		}

		return fieldBlocks;
	}

	@Override
	public boolean isInField(IFieldInteraction projector, Vector3 position)
	{
		Vector3 posScale = projector.getPositiveScale().clone();
		Vector3 negScale = projector.getNegativeScale().clone();

		int xStretch = posScale.xi() + negScale.xi();
		int yStretch = posScale.yi() + negScale.yi();
		int zStretch = posScale.zi() + negScale.zi();

		Vector3 projectorPos = new Vector3((TileEntity) projector);
		projectorPos.add(projector.getTranslation());
		projectorPos.add(new Vector3(0, -negScale.yi() + 1, 0));

		Vector3 relativePosition = position.clone().subtract(projectorPos);
		relativePosition.rotate(-projector.getRotationYaw(), -projector.getRotationPitch());

		Cuboid region = new Cuboid(negScale.scale(-1), posScale);

		if (region.isIn(relativePosition) && relativePosition.y > 0)
		{
			if ((1 - (Math.abs(relativePosition.x) / xStretch) - (Math.abs(relativePosition.z) / zStretch) > relativePosition.y / yStretch))
			{
				return true;
			}
		}

		return false;
	}

	@SideOnly(Side.CLIENT)
	@Override
	public void render(IProjector projector, double x, double y, double z, float f, long ticks)
	{
		Tessellator tessellator = Tessellator.instance;

		GL11.glPushMatrix();
		GL11.glRotatef(180, 0, 0, 1);

		float height = 0.5f;
		float width = 0.3f;
		int uvMaxX = 2;
		int uvMaxY = 2;
		Vector3 translation = new Vector3(0, -0.4, 0);
		tessellator.startDrawing(6);
		tessellator.setColorRGBA(72, 198, 255, 255);
		tessellator.addVertexWithUV(0 + translation.x, 0 + translation.y, 0 + translation.z, 0, 0);
		tessellator.addVertexWithUV(-width + translation.x, height + translation.y, -width + translation.z, -uvMaxX, -uvMaxY);
		tessellator.addVertexWithUV(-width + translation.x, height + translation.y, width + translation.z, -uvMaxX, uvMaxY);
		tessellator.addVertexWithUV(width + translation.x, height + translation.y, width + translation.z, uvMaxX, uvMaxY);
		tessellator.addVertexWithUV(width + translation.x, height + translation.y, -width + translation.z, uvMaxX, -uvMaxY);
		tessellator.addVertexWithUV(-width + translation.x, height + translation.y, -width + translation.z, -uvMaxX, -uvMaxY);
		tessellator.draw();
		GL11.glPopMatrix();
	}
}

File: src/main/scala/mffs/field/mode/ItemModePyramid.scala
package mffs.field.mode

import java.util.{HashSet, Set}

class ItemModePyramid extends ItemMode
{
  private val step = 1

	def getExteriorPoints(projector: IFieldMatrix): Set[Vector3d] =
  {
	  val fieldBlocks = new HashSet[Vector3d]

    val posScale = projector.getPositiveScale
    val negScale = projector.getNegativeScale

    var xSize = Math.max(posScale.x, Math.abs(negScale.x))
    var zSize = Math.max(posScale.z, Math.abs(negScale.z))
    val ySize = Math.max(posScale.y, Math.abs(negScale.y)).asInstanceOf[Int]

    val initX = xSize.asInstanceOf[Int]
    val initZ = xSize.asInstanceOf[Int]

    val xDecr = xSize / (ySize * 2)
    val zDecr = zSize / (ySize * 2)

    //Create pyramid
    for (y <- -ySize to ySize)
    {
      for (x <- -initX to initX; z <- -initZ to initZ)
      {
		  if (Math.abs(x) == Math.round(xSize) && Math.abs(z) <= Math.round(zSize)) {
			  fieldBlocks.add(new Vector3d(x, y, z))
		  }
		  else if (Math.abs(z) == Math.round(zSize) && Math.abs(x) <= Math.round(xSize))
			  fieldBlocks.add(new Vector3d(x, y, z))
		  else if (y == -ySize)
			  fieldBlocks.add(new Vector3d(x, y, z))
      }

      xSize -= xDecr
      zSize -= zDecr
    }

    return fieldBlocks
  }

	def getInteriorPoints(projector: IFieldMatrix): Set[Vector3d] =
  {
	  val fieldBlocks: Set[Vector3d] = new HashSet[Vector3d]
	  val posScale: Vector3d = projector.getPositiveScale
	  val negScale: Vector3d = projector.getNegativeScale
    val xStretch: Int = posScale.xi + negScale.xi
    val yStretch: Int = posScale.yi + negScale.yi
    val zStretch: Int = posScale.zi + negScale.zi
	  val translation = new Vector3d(0, -0.4, 0)

    for (x <- -xStretch to xStretch by step; y <- 0 to yStretch by step; z <- -zStretch to zStretch by step)
    {
		val position = new Vector3d(x, y, z) + translation

		if (isInField(projector, position + new Vector3d(projector.asInstanceOf[TileEntity])))
      {
        fieldBlocks.add(position)
      }
    }
    return fieldBlocks
  }

	override def isInField(projector: IFieldMatrix, position: Vector3d): Boolean =
  {
	  val posScale: Vector3d = projector.getPositiveScale.clone
	  val negScale: Vector3d = projector.getNegativeScale.clone
    val xStretch: Int = posScale.xi + negScale.xi
    val yStretch: Int = posScale.yi + negScale.yi
    val zStretch: Int = posScale.zi + negScale.zi
	  val projectorPos: Vector3d = new Vector3d(projector.asInstanceOf[TileEntity])
    projectorPos.add(projector.getTranslation)
	  projectorPos.add(new Vector3d(0, -negScale.yi + 1, 0))
	  val relativePosition: Vector3d = position.clone.subtract(projectorPos)
    relativePosition.transform(new EulerAngle(-projector.getRotationYaw, -projector.getRotationPitch, 0))
    val region: Cuboid = new Cuboid(-negScale, posScale)

    if (region.intersects(relativePosition) && relativePosition.y > 0)
    {
      if ((1 - (Math.abs(relativePosition.x) / xStretch) - (Math.abs(relativePosition.z) / zStretch) > relativePosition.y / yStretch))
      {
        return true
      }
    }

    return false
  }

  @SideOnly(Side.CLIENT) override def render(projector: IProjector, x: Double, y: Double, z: Double, f: Float, ticks: Long)
  {
    val tessellator = Tessellator.instance
    GL11.glPushMatrix
    GL11.glRotatef(180, 0, 0, 1)
    val height: Float = 0.5f
    val width: Float = 0.3f
    val uvMaxX: Int = 2
    val uvMaxY: Int = 2
	  val translation: Vector3d = new Vector3d(0, -0.4, 0)
    tessellator.startDrawing(6)
    tessellator.setColorRGBA(72, 198, 255, 255)
    tessellator.addVertexWithUV(0 + translation.x, 0 + translation.y, 0 + translation.z, 0, 0)
    tessellator.addVertexWithUV(-width + translation.x, height + translation.y, -width + translation.z, -uvMaxX, -uvMaxY)
    tessellator.addVertexWithUV(-width + translation.x, height + translation.y, width + translation.z, -uvMaxX, uvMaxY)
    tessellator.addVertexWithUV(width + translation.x, height + translation.y, width + translation.z, uvMaxX, uvMaxY)
    tessellator.addVertexWithUV(width + translation.x, height + translation.y, -width + translation.z, uvMaxX, -uvMaxY)
    tessellator.addVertexWithUV(-width + translation.x, height + translation.y, -width + translation.z, -uvMaxX, -uvMaxY)
    tessellator.draw
    GL11.glPopMatrix
  }
}